docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+
+  `Prefetch` internals:
+
+  - ensure `end` is not great than `f.size`
+
+  - convert ranges to page ids
+- get new page ids (not in f.pages)
+  - if have not new page ids (have been fetched)
+  - return a slice of mmaped file (f.data[start:end])
+  
+- add new page into `f.pendingPages`
+  - increase `f.pendingReaders`
+  - return `f.pendingChan`
+* Range: 
+
+  `Range` returns a slice of mmaped file (`f.data[start:end]`). 
+
+  It will not try to fetch data from obj store, and return an error if some pages in the range have not been fetched.
+
+  `Range` internals:
+
+  - convert range to page ids  
+  - if all page ids have been fetched
+    - return a slice of mmaped file
+
+  * otherwise return an error
+
+* CloseMmap: `CloseMmap` unmmaps `f.data` .
+
+* CloseFile: `CloseFile` closes  `f.file`. 
+
+* Close: `Close` closes `f.file`, unmmaps `f.data` and set `f.pages` to nil.
+
+### Range to pages
+
+Convert a range to pages:
+
+```
+const pageSize = 4 * 1024  // 4KiB
+
+minPageID := math.Floor( float(start) / float(pageSize) )
+maxPageID := math.Ceil( float(end) / float(pageSize) )
+
+pages := roaring.NewBitmap()
+pages.AddRange(minPageID, maxPageID)
+```
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next  time to fetch. So that we can combine multiple queries to ones.
+
+When fetching combined  pages,
+if an error occurs, BucketFile will send the error to callers via the `pendingChan` channel,
+if no error occurs, BucketFile will close the `pendingChan` channel.
+
+If callers receive an error from the channel, then prefetch failed. 
+If callers receive nil (zero value of error, after channel closed) from the channel, then prefetch done. 
+
+Each time to fetch pending pages，BucketFile will reset  `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory object
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into memory object for sequential and  random accesses.  
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses very small memory. 
+
+For inactive IndexReaders, if the inactive duration (`now.Sub(lastQueryTime)`) greats than a configured duration (e.g. 15 mins), then IndexReader will remove in-memory objects and close file descriptor and mmap in IndexFile.  This will reduce memory usage.
```

Comment:
Who will actually track this time?

Reply:
`IndexReaders` and `ChunksReader`

Reply:
there will be separate tracking for those two?

Reply:
Update: `BucketFile` will track the time.

IndexReaders hold decoded `postings offsets` and `label offsets` in memory. 
so IndexReaders need to track the time.
if one IndexReader is inactive, it will remove in-memory `postings offsets` and `label offsets`.

ChunksReader no need to track.